### ISSUE TITLE
[bug]: `LinearFeeFunction` no check of the budget if conf_target <=1

### ISSUE THREAD

**Issue by feelancer21** [lightningnetwork/lnd#9102]
If `conf_target<=1` we return a `LinearFeeFunction` immediately. But the `maxFeeRate` can be derived from the budget and be lower than the current fee estimate for 1 block. Imho the current fee estimation should be checked before returning the fee function,

https://github.com/lightningnetwork/lnd/blob/e8c5e7d5ce4b79117ef9ab497745fea3a1bffb36/sweep/fee_function.go#L117-L131

**Comment by Roasbeef** [comment]
> ut the maxFeeRate can be derived from the budget and be lower than the current fee estimate for 1 block. Imho the current fee estimation should be checked before returning the fee function,

The whole point of `maxFeeRate` is that it's an absolute limit on the fee function. If we allow the next block estimate to supersede it, then`maxFeeRate` does nothing.

**Comment by feelancer21** [comment]
Hello Roasbeef, 

For the case `confTarget>1`, the **Error(“fee rate delta is zero”)` is returned if the budget is insufficient. In the case `confTarget<=1`, however, no error is returned but the FeeFunction, which may also return a fee rate from the budget that is too low. 

But I think this case is intercepted elsewhere. 
https://github.com/lightningnetwork/lnd/blob/750770e1904dd7db54b2cf1aa510d5e92308e290/sweep/fee_bumper.go#L499-L503

So I think everything is fine.

**Comment by yyforyongyu** [comment]
> But the maxFeeRate can be derived from the budget and be lower than the current fee estimate for 1 block.

Yes it can be lower, and sometimes we want it to be lower to save the cost. If that's not wanted, we should increase the budget instead of trying to change the function here. The budget can be increased using multiple ways, either [via config](https://github.com/lightningnetwork/lnd/tree/master/sweep#sweeping-outputs-from-a-force-close-transaction) or [via cli](https://docs.lightning.engineering/lightning-network-tools/lnd/sweeper#hands-on-managing-sweeps).

**Comment by feelancer21** [comment]
@yyforyongyu 
Perhaps another example to make my point clearer. We have a maxFeeRate of 100 sat/vb. But the FeeEstimator returns 200 sat/vb for all possible confTargets. If `confTarget>1` in this case `Error(“fee rate delta is zero”)` is returned. However, if `confTarget<=1`, a FeeFunction is always returned, which returns a maxFeeRate of 100 sat/vb. This means that a sweep tx may be created that would be in the budget but would not be successful for confirmation unless this is caught elsewhere. I'm not quite sure about the latter. That's why I opened the issue.
Perhaps it is even intentional that we want to put tx in the mempool that have no change of success in the near future.

**Comment by yyforyongyu** [comment]
> Perhaps it is even intentional that we want to put tx in the mempool that have no change of success in the near future.

Exactly. The whole budget-based sweeper takes the idea that, if a user decides this is the max fee he/she is willing to pay, the sweeper will respect that, even if it means the tx cannot be confirmed at the next block.

Not all sweeping txns are time-sensitive, e.g., the `to_local` output, so it makes sense for users to wait for more blocks if the current mempool is congested. On the other hand, even when an output is time-sensitive, users may choose not to sweep it due to economic reasons. To express this economic preference, we use `budget`.

Imagine an HTLC with a small value, sth like 1000 sats, given the current feerate of 200 sats/vb, it won't make sense to sweep it, since it means the user may even need to use a wallet utxo, and the end profit is negative.

Furthermore, we don't want to rely on the fee estimator too much, especially when it comes to estimating the likelihood of getting the tx confirmed in the next block, my observation is the fee estimator is not a reliable source. We may need to build a more powerful fee estimator here, but it's not a simple business to predict what feerate the miners would use to package txns, tho it may be more straightforward once the package mempool is in, future stories tho.

**Comment by feelancer21** [comment]
> Exactly. The whole budget-based sweeper takes the idea that, if a user decides this is the max fee he/she is willing to pay, the sweeper will respect that, even if it means the tx cannot be confirmed at the next block.

Ok. Last question. Suppose all my utxos in the onchain wallet are now blocked by sweeps from htlcs that are outside the wallet. Now suddenly there is a force close occured with a high-value htlc that absolutely must be saved. Will one of the blocked utxo be replaced via RBF?

**Comment by ziggie1984** [comment]
I think @feelancer21 has a valid point here:

Following scenario:

Sweeper situation:

1 wallet utxo: 100k sats
1 outgoing sweep HTLC: 5000 sats => budget *2 => 10k sats
ctlv_delta 144 => expiry of the incoming is 144 blocks away
to_local csv >500 blocks
ChainFee: 1000 sats/vbyte  => For Conftarget 1
                   500 sats/vbyte => Conftarget 144
MinRelayFee: 10 sats/vbyte

Default Budget Ratio is 0.5 => budget is 2500 sats => maxfeerate
`HtlcTimeoutWeightConfirmed = 666` => ((500 * 250 sats/kweight *666 weight )/1000 = 83250 sats)
=> Channel gets force-closed

- MinRelay Check succeeds, outgoing htlc could pay the min_relay fee.
- Broadcasting the sweep fails I guess here:
https://github.com/lightningnetwork/lnd/blob/edd9ade7e555aeea675f00dcd50e10130dab1f3d/sweep/fee_bumper.go#L498-L504
- Basically we would always fail as long as we are not hitting the conf_target <= 1 (assuming fee-estiamtion does not change)
- Now when hitting the conf_target <= 1 case, we will broadcast the sweep with a max fee of 10k sats => blocking the only wallet input, although there is no real chance this sweep will confirm soonish.


1. Maybe we should not attempt the sweep in this case ?
2. Maybe we need some kind of priority system where we would use Unconfirmed wallet inputs to sweep more important htlcs.

**Comment by Roasbeef** [comment]
> 1 outgoing sweep HTLC: 5000 sats => budget *2 => 10k sats
> HtlcTimeoutWeightConfirmed = 666 => ((500 * 250 sats/kweight *666 weight )/1000 = 83250 sats)

In this case, if we sweep at that level then pretty much all the funds would to to chain fees, and we'd burn most of our existing UTXO at that point. In the past we used the concept of negative yield to guard against this, and now we have the budget for this exact reason. 

> Not broadcasting the sweep if the acutal mempool condition did change might be a solution but I don't think its a proper solution to this problem and we should aim for the replaceable function using 0-conf wallet inputs.

Sweeping in the conditions you laid out above would just lose funds for the user, I don't think we want to do that. 


> Now when hitting the conf_target <= 1 case, we will broadcast the sweep with a max fee of 5k sats => blocking the only wallet input, although there is no real chance this sweep will confirm soonish.

We can certainly add more and more policies to attempt to get to an ideal solution, but I think ultimately we'll have to acknowledge that there're degenerate cases that may lead to a sub-optimal outcome.

**Comment by ziggie1984** [comment]
I think as soon as we have the option of also cancelling back non-dust-htlcs (up to a certain amount in total) which I am working on, we can return back here.

The main question is the following I wanted to point out:

We need to balance blocking a wallet input compared to the probability of a sweep confirming. That's why I mentioned above we should probably attempt the sweep as we do now, but make sure if a new more valuable HTLC is queued to be swept we need to take the used wallet input of the unlikely sweep-tx, and use it to sweep the high amount htlc.
